source/client-side-encryption/client-side-encryption.rst

```diff
@@ -790,10 +898,215 @@ MUST be distinguished in some way (e.g. exception type) to make it
 easier for users to distinguish when a command fails due to client side
 encryption.
 
+
+.. index:: csfle
+.. _Enabling csfle:
+
+Enabling Command Marking with the csfle_ Library
+=================================================
+
+The MongoDB Enterprise distribution includes a dynamic library named
+``mongo_csfle_v1`` (with the appropriate file extension or filename suffix for
+the host platform). This library will be loaded by libmongocrypt_ when the
+``mongocrypt_init`` function is invoked
+`(from the libmongocrypt C API) <lmc-c-api_>`_ based on the search criteria that
+are provided by the driver.
+
+libmongocrypt_ allows the driver to specify an arbitrary list of directory paths
+in which to search for the csfle_ dynamic library. The user may specify csfle_
+search behavior by specifying options in extraOptions_: |opt-paths| and
+|opt-path-override|.
+
+.. note::
+
+   The driver MUST NOT manipulate or do any validation on the csfle_ path
+   options provided in extraOptions_. They should be passed through to
+   libmongocrypt_ unchanged.
+
+
+.. _search path:
+.. _search paths:
+
+Setting Search Paths
+--------------------
+
+The driver MUST append csfle_ search paths to the `libmongocrypt_handle`. For
+each path :math:`P` in |opt-paths|, append :math:`P` to the search paths on the
+`libmongocrypt_handle`, maintaining the order of the user-provided array.
+
+
+.. rubric:: Explaination
+
+The |opt-paths| array allows the user to customize the location of the csfle_
+library that they want to use for their application. For example, a user may
+have bundled a csfle_ library with their application and knows where in the
+bundle that library is located.
+
+.. seealso:: `Path Resolution Behavior`_
+
+
+.. _override path:
+
+Overriding the csfle_ Library Path
+----------------------------------
+
+If |opt-path-override| was specified by the user, the driver MUST set the
+csfle_ path override on the `libmongocrypt_handle`.
+
+.. note::
+
+   If a path override is set on a `libmongocrypt_handle` and libmongocrypt_
+   fails to load csfle_ from that filepath, then that will result in a
+   hard-error when initializing libmongocrypt_.
+
+
+Path Resolution Behavior
+------------------------
+
+.. |---| unicode:: U+2014
+
+Drivers should include and note the following information regarding the behavior
+of csfle_ path options in extraOptions_:
+
+- If used, the `override path`_ must be given as a path to the csfle_ dynamic
+  library file *itself*, and not simply the directory that contains it.
+
+- If the `override path`_ is specified, the `search paths`_ will be ignored.
+
+- For csfle_ `search paths`_ and the `override path`_, if the given path is a
+  relative path and the first path component is the literal string
+  :ts:`"$ORIGIN"`, the :ts:`"$ORIGIN"` component will be replaced by the
+  absolute path to the directory containing the libmongocrypt_ library that is
+  performing the csfle_ search. This behavior mimics the ``$ORIGIN`` behavior of
```

Comment:
Possibly a different solution in that case, but it may depend on the driver/appliation. The general idea is to allow applications to sit beside their dynamic libraries and enforce the dynamic linker to prefer the once that it is bundled with the application rather than the ones on the system. Of course, on systems that support `RPATH` they can just set their app's `RPATH` to `$ORIGIN/../lib`, but not all languages will be able to set the `RPATH`, in which can `LD_LIBRARY_PATH`.